models/exposure.py

```diff
@@ -1,7 +1,238 @@
 
-from models.base import Base
+import inspect
+from collections import defaultdict
 
+import numpy as np
+
+import sqlalchemy as sa
+from sqlalchemy import orm
+from sqlalchemy.dialects.postgresql import JSONB
+
+from astropy.coordinates import SkyCoord
+
+from models.base import Base, SpatiallyIndexed
+import models.instruments
+
+from pipeline.utils import normalize_header_key
+
+class CCD_Data:
+    """
+    A helper class that lazy loads the CCD data from the database.
+    When requesting one of the CCD IDs it will fetch that data from
+    disk and store it in memory.
+    To clear the memory cache, call the clear_cache() method.
+    """
+    def __init__(self, filename, inst):
+        self.filename = filename
+        self.instrument_object = inst
+        self._data = defaultdict(lambda: None)
+
+    def __getitem__(self, ccd_id):
+        if self._data[ccd_id] is None:
+            self._data[ccd_id] = self.instrument_object.load(self.filename, ccd_id)
+        return self._data[ccd_id]
+
+    def __setitem__(self, ccd_id, value):
+        self._data[ccd_id] = value
+
+    def clear_cache(self):
+        self._data = defaultdict(lambda: None)
+
+
+class Exposure(Base, SpatiallyIndexed):
 
-class Exposure(Base):
     __tablename__ = "exposures"
 
+    header = sa.Column(JSONB, nullable=False, default={}, doc="Header of the raw exposure. ")
+
+    filename = sa.Column(sa.Text, nullable=False, index=True, unique=True, doc="Filename for raw exposure. ")
+
+    mjd = sa.Column(
+        sa.Double,
+        nullable=False,
+        index=True,
+        doc="Modified Julian date of the exposure. (MJD=JD-2400000.5)"
+    )
+
+    exp_time = sa.Column(sa.Float, nullable=False, index=True, doc="Exposure time in seconds")
+
+    filter = sa.Column(sa.Text, nullable=False, index=True, doc="Filter name")
+
+    instrument = sa.Column(sa.Text, nullable=False, index=True, doc="Instrument name")
```

Comment:
I noticed in CCDData that you pass something called an instrument_object.

Is the idea that somewhere we'll have an Instrument class, and then we'll make a subclass for each actual instrument, and then this field is going to be an index into a mapping of name->subclass?

Reply:
A reason to have an instrument table instead of just classes is so that things like num_ccds, width, height can go into that table instead of being repeated every row of the exposures table.

Reply:
That's a good point. At some point we will need to match the instrument row to the instrument class (where the code to do things lives). 
If we just assume each Exposure's instrument name is matched to a hard-coded instrument class, then we can just load an object of that class (or the pertinent info like width/height) whenever the Exposure is loaded. But it doesn't enforce that each instrument name is indeed matched to an instrument class. 
On the other hand, if we have an instrument table we should still allow the user to input the instrument as a string and match that internally using a foreign key. 
I think I can come up with something like that.